商品購入機能

takonushi opened the pull request:
# What
　商品購入機能実装

# Why
　商品購入機能実装のため

・必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
　https://gyazo.com/b88d341a7273dc2cb0210598a4edc01a
・入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
　https://gyazo.com/57cd83d4e9d53fdece97acce006c44bd
・ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
　https://gyazo.com/10e8d0e063d649890cb573b917fd39c7
・ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
　https://gyazo.com/74a095cbb6c38ce58b2a43541f4e130b
・ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインペ　ージに遷移する動画
　https://gyazo.com/579f128aa2949c458a1f4cbaac524bf9
・売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
・売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
　https://gyazo.com/22d4f6eadd01955a9114402d0a018708 　※上記2つ分
・ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
　https://gyazo.com/e9760cf4002dda48babce7577cf8d2cf
・ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
　https://gyazo.com/e474e7eaa98e1a31bc48baeb9eeb7b45
・テスト結果の画像
　https://gyazo.com/b5a7410653d260560a2b9bc34626bbb0

takonushi committed:
orderモデル・テーブル作成

・orderモデル作成
・ordersテーブル作成

takonushi committed:
orders関連の設定

・ordersコントローラの作成
・ルーティングの追記(itemsのネスト構造)
・views/orders/index.html.erbを設置
・ordersコントローラーにindex/createアクションの定義

takonushi committed:
indexアクション修正、htmlファイルの修正

・indexアクションの修正
・items/show.html.erb修正
・orders/index.html.erb修正

takonushi committed:
クレジットカード決済機能の実装

・card.jsによるカード情報トークン化、サーバーサイドへの情報送信まで実装
・payjpの導入
・秘密鍵、公開鍵は環境変数化ずみ
・orders/index.html.erbへ金額情報をjavascriptへ渡す記述を追記
・テストコードについては未実装
・実際にはsendingsテーブルへもデータ更新を行う動作。現時点でのフォームは　
　orderテーブルのみへの更新(create時に一部エラーも発生中)

takonushi committed:
sendingテーブル作成、アソシエーション追記

・sendingテーブル作成
・各モデルへのアソシエーション追記(order / sendingに対して)

takonushi committed:
Formオブジェクト実装前の暫定実装

・orderテーブルとsendingテーブルそれぞれにデータが書き込まれるように修正
　(Formオブジェクト実装前のため、暫定的な設定)

takonushi committed:
複数フォームへの実装(バリデーション確認等未実施)

・order/sendingテーブル両方に登録がされること確認済み
・payjpまでデータを渡せること確認済み
・render処理でエラーが出ているのとバリデーション周りは未確認

takonushi committed:
購入時のエラーハンドル・未購入,購入済みのビュー実装

・購入時にエラーハンドルされる様に実装
・購入済みの時のみsoldout,未購入の時のみ購入ボタンが表示される様に実装

takonushi committed:
バリデーションの設定調整、単体コードの実装

・購入済み状態では削除・編集出来ない様にアクションの制限を実装
・バリデーションの設定調整
・購入済み・未購入状態におけるビューを実装
・Formオブジェクト(order_sending)に対する単体テストコードを実装

takonushi committed:
Rubocop実行後

div-reviewer commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,38 @@
+  before_action :authenticate_user!, only: [:index, :create]
+
+  def index
+    @item = Item.find(params[:item_id])
```
```ruby  
  @item = Item.find(params[:item_id])
```
こちらの記述が複数のメソッドで使用されているかと思われます。
よってprivate以下にメソッド化し、before_actionで呼び出しましょう。可読性を上げるためです。
複数回使用する変数をメソッド化することによってまず可読性が上がります。また保守（修作業など）が発生した際も手を加えやすくなります。

```ruby
【実装イメージ】
before_action :set_item, ~~~

def set_item
 @item = Item.find(params[:item_id])
end
```

>参考カリキュラム
フェーズ：応用
タイトル：Ruby on Rails（PicTweet）〜ツイートの詳細を表示しよう〜
目次：同じコードをメソッドにまとめよう

div-reviewer commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,38 @@
+  def pay_item
+    Payjp.api_key = ENV['PAYJP_SECRET_KEY']
+    Payjp::Charge.create(
+      amount: order_params[:price],  # 商品の値段
```
```ruby
amount: order_params[:price],  # 商品の値段
```

order_params内からprice: params[:price]を削除し、
amount: @item.priceに書き換えましょう。

amountの設定はorder_params〜でないといけない訳ではない為、params内で再作成しなくても良い為です。

@itemがcreate内で出来上がっていることが前提となりますので、先にbeforeアクションでfindをまとめる動きを実装しましょう。

div-reviewer commented on spec/models/order_sending_spec.rb:
```diff
@@ -0,0 +1,85 @@
+        expect(@order_sending.errors.full_messages).to include("Telephone number can't be blank")
+      end
+
+      it '電話番号の桁数が10桁以上、11桁以内の半角数値以外では登録ができない' do
+        @order_sending.telephone_number = '080-1111-2222'
+        @order_sending.valid?
+        expect(@order_sending.errors.full_messages).to include('Telephone number is invalid. Telephone number must be 10 or 11 single-byte numbers')
+      end
```
電話番号の半角数値以外で登録できないテストができていますね。
電話番号の桁が半角数値かつ9桁、半角数値かつ12桁以上でも登録できないテストを追加しましょう。

以下のバリデーションを確認するため。
```ruby
validates :telephone_number,
              format: { with: /\A\d{10,11}\z/, message: 'is invalid. Telephone number must be 10 or 11 single-byte numbers' }
```

div-reviewer commented on app/javascript/card.js:
```diff
@@ -0,0 +1,40 @@
+      Payjp.createToken(card, (status, response) => {
+        if (status == 200) {
+          const token = response.id;
+          const price = JSON.parse(document.getElementById('price_info').dataset.json);
```
```ruby
const price = JSON.parse(document.getElementById('price_info').dataset.json);
 const priceObj = `<input value=${price} name='price' type="hidden"> `;
renderDom.insertAdjacentHTML("beforeend", priceObj);
```

購入機能では、金額を再び保存するのではなく、商品出品の際に保存したデータを取得して使用しますので上記の記述は不要となります。よって削除しましょう。

div-reviewer commented on app/models/order.rb:
```diff
@@ -0,0 +1,7 @@
+class Order < ApplicationRecord
+  attr_accessor :token, :price
```
```ruby
attr_accessor :token, :price
```

購入管理モデルでtoken、priceは特に使用しておりませんのでこちらは不要かと思われます。
よって削除しましょう。

div-reviewer reviewed (commented):
takonushi様、学習お疲れさまです
お忙しい中購入機能の実装と、プルリクエストの提出ありがとうございました。

フォームオブジェクトのバリデーションでwith_optionsが使用できたり、コントローラーの条件もきれいに書けているかと思われます。素晴らしいです。

今回5点コメントさせて頂きましたのでご確認をお願いいたします。
質問や疑問点がございましたらコネクトかチャットメンターにお問い合わせ下さい。

どうぞ宜しくお願い致します。

takonushi committed:
各種項目の修正

・同一記述のメソッド化
・orderモデルの不要な記述(attr_accessor)を削除
・カードの金額(price)情報の引き渡し方法を変更。(Formからの引き渡しではなく、Itemテーブルからの呼び出し)
・上記に伴うorders/index.html.erbのprice引き渡し用の記載の削除、card.jsのprice引き渡し用の情報を削除
・テストコード(order_sending_spec.rb)のtelephone_numberに対して9桁以下の半角数値に対するテストと12桁以上の半角数値に対するテスト項目を追加

takonushi commented:
ご指摘ありがとうございます。いただいた5点の内容に関して修正を行いましたのでご確認のほどお願いいたします。

・同一記述のメソッド化(set_item)
・orderモデルの不要な記述(attr_accessor)を削除
・カードの金額(price)情報の引き渡し方法を変更。(ビューから引き渡しではなく、Itemテーブルからの呼び出し)
・上記に伴うorders/index.html.erbのprice引き渡し用の記載の削除、card.jsのprice引き渡し用の情報を削除
・テストコード(order_sending_spec.rb)のtelephone_numberに対して9桁以下の半角数値に対するテストと12桁以上の半角数値に対するテスト項目を追加

テスト結果
https://gyazo.com/7ccf42cf8eb8623e1911df14ba39f6d2

greenichi commented on app/models/order_sending.rb:
```diff
@@ -0,0 +1,24 @@
+class OrderSending
+  include ActiveModel::Model
+  attr_accessor :postal_code, :prefecture_id, :city, :house_number, :building_name, :telephone_number, :order_id, :user_id,
```
```
:order_id
:price
```
こちらは削除しましょう。

理由：
order_idはsaveメソッド内でcreateするためです。
priceの情報はOrderSendingの属性として取り扱う必要がないためです。

こちらご修正後のレビューの再提出は不要となります！

greenichi reviewed (commented):
レビュー依頼ありがとうございます！

修正点を確認いたしました。

**LGTM**とさせていただきます！
実装お疲れ様でした！

但し、マージ前に1つ確認いただきたい事項がございます。
1点コメントしておりますので、そちらの修正をお願いいたします。
修正後に再度レビューをご提出いただく必要はございません！

修正後、こちらのブランチは忘れずにMergeいただき、
ローカルのMasterにPullした後に次の実装のためのブランチを切って次の課題に取り組みましょう！
購入機能実装まで、本当にお疲れさまでした。
引き続き、開発頑張っていきましょう！応援しております。

takonushi committed:
不要な記述の削除
